File: leetcode/002-add-two-numbers.cxx
// https://leetcode.com/problems/add-two-numbers

// struct ListNode {
//   int val;
//   ListNode *next;
//   ListNode(int x) : val(x), next(NULL) {}
// };

class Solution {
public:
  ListNode *addTwoNumbers(ListNode *xs, ListNode *ys) {
    if (xs != NULL and ys != NULL) {
      int mval = xs->val + ys->val;
      ListNode *next = calcNext(mval >= 10, addTwoNumbers(xs->next, ys->next));
      return cons(mval % 10, next);
    } else if (xs == NULL) {
      return ys;
    } else if (ys == NULL) {
      return xs;
    } else {
      return NULL;
    }
  }

  inline ListNode *calcNext(bool carry, ListNode *mnext) {
    if (carry) {
      return addTwoNumbers(mnext, new ListNode(1));
    } else {
      return mnext;
    }
  }

  inline ListNode *cons(int headVal, ListNode *tail) {
    ListNode *node = new ListNode(headVal);
    node->next = tail;
    return node;
  }
};

File: leetcode/002-add-two-numbers.scala
// https://leetcode.com/problems/add-two-numbers

/**
  * class ListNode(
  *   var x: Int = 0,
  *   var next: ListNode = null
  * )
  */

object Solution {

  def addTwoNumbers(xs: ListNode, ys: ListNode): ListNode =
    if (xs eq null) ys else if (ys eq null) xs else {
      val n = xs.x + ys.x
      val ns = addTwoNumbers(xs.next, ys.next)
      if (n < 10) cons(n)(ns) else cons(n % 10)(addTwoNumbers(ns, new ListNode(1)))
    }

  def cons(x: Int) = (xs: ListNode) => {
    val n = new ListNode(x)
    n.next = xs
    n
  }

}
